data_channel.c

```diff
@@ -515,17 +515,39 @@ exchange_params2(struct data_channel *data_channel)
       param->value[sizeof(param->value) - 1] = 0;
     }
 
+    /* for ADF */
+    param = get_scan_param_by_id(data_channel, 'Z');
+    assert(param);
+    sscanf(param->value, "%ld,%ld,%ld,%ld", &adf_startx, &adf_starty, &adf_width, &adf_height);
+
     param = get_scan_param_by_id(data_channel, 'A');
     assert(param);
-    sprintf(param->value, "0,0,%ld,%ld", recv_params[4], recv_params[6]);
+    sscanf(param->value, "%ld,%ld,%ld,%ld", &startx, &starty, &width, &height);
+
+    if ((recv_params[6] <= 0) && (adf_height > 0)) {
```

Comment:
I will add some comments. You assumed right, if Z parameter isn't set, adf_height is 0.
The logic is : did we receive 0 as endy from scanner ? In that case, we assume that we are scanning from adf (my adf doesn't send back the height of the page, I receive it only when scanning from flatbed), and we use Z parameters if set. Otherwise, use A parameters if set. Otherwise use parameters received from scanner.

Reply:
By the way, it corrects #3